### PR TITLE
Record ids of removed virtual functions

### PR DESCRIPTION
**Pull request by mgudemann**
Allows for bounding the unwinding of virtual functions in symex phase.

### CODE REVIEW

**Review by smowton — COMMENTED** [review]
Rather than adding extra side-effects to RVF wouldn't it be simpler to just make a pass over the goto_functionst beforehand and record the virtual function callsites?

File: src/goto-analyzer/goto_analyzer_parse_options.cpp [hunk 1]
```diff
@@ -373,7 +373,8 @@ bool goto_analyzer_parse_optionst::process_goto_program(
     remove_function_pointers(
       get_message_handler(), goto_model, cmdline.isset("pointer-check"));
     // Java virtual functions -> explicit dispatch tables:
-    remove_virtual_functions(goto_model);
+    std::vector<irep_idt> virtual_functions;
```
Comment (thk123) [inline src/goto-analyzer/goto_analyzer_parse_options.cpp]: As opposed to adding unnecessary vectors that are discarded, consider overloading ` remove_virtual_functions` to create its own empty vector.

**Comment by mgudemann** [comment]
@smowton this was the original approach in https://github.com/diffblue/test-gen/pull/1087 but in the test-case `Inner3.b` has no call site and was missed therefore. `remove_virtual_functions` treats all virtual functions, so I chose to simply record the functions ids it replaces.

**Comment by smowton** [comment]
So the out-of-line code you've got there says
```
for(const auto &goto_function : goto_model.goto_functions.function_map)
  {
    const goto_programt &goto_program=goto_function.second.body;

    for(const auto &target : goto_program.instructions)
    {
      if(target.is_function_call())
      {
        const code_function_callt &code=
          to_code_function_call(target.code);

        if(code.function().id()==ID_virtual_function)
        {
          const irep_idt &function_id=code.function().get(ID_identifier);
          virtual_function_ids.push_back(function_id);
```

And remove-virtuals says:

```
Forall_goto_program_instructions(target, goto_program)
    if(target->is_function_call())
    {
      const code_function_callt &code=
        to_code_function_call(target->code);

      if(code.function().id()==ID_virtual_function)
      {
        remove_virtual_function(goto_program, target);

...

const code_function_callt &code=
    to_code_function_call(target->code);

  const auto &vcall_source_loc=target->source_location;

  // record the virtual functions that are removed
  virtual_functions.push_back(target->function);
```

Looks like the only difference is recording the callee (TG-16 version) vs. the caller (this version). Couldn't you achieve the same effect by simply changing `const irep_idt &function_id=code.function().get(ID_identifier);` to `const irep_idt &function_id=target.function`?

**Comment by martin-cs** [comment]
At the risk of being Unhelpful, is a random vector of strings doesn't seem like the best way of storing this information.  How is it used?  Would it be better as a comment field in the program and / or an annotation in the symbol table?

File: src/goto-programs/remove_virtual_functions.h [hunk 2]
```diff
@@ -21,8 +21,14 @@ Date: April 2016
 void remove_virtual_functions(
   goto_modelt &goto_model);
 
+// record removed virtual function ids
+void remove_virtual_functions(
+  goto_modelt &goto_model,
+  std::vector<irep_idt> &virtual_functions);
+
```
Comment (martin-cs) [inline src/goto-programs/remove_virtual_functions.h:28]: Adding a new signature is a good way of extending the API without breaking it.

File: src/goto-programs/remove_virtual_functions.h [hunk 3]
```diff
@@ -21,8 +21,14 @@ Date: April 2016
...
 void remove_virtual_functions(
   const symbol_tablet &symbol_table,
-  goto_functionst &goto_functions);
+  goto_functionst &goto_functions,
+  std::vector<irep_idt> &virtual_functions);
```
Comment (martin-cs) [inline src/goto-programs/remove_virtual_functions.h:33]: Any chance we could have it here as well?

**Review by martin-cs — APPROVED** [review]
Appears to be correctly implemented.  I'm not sure the interface is necessarily the best way of passing the information around but that's not an answerable question just looking at this PR.

**Comment by tautschnig** [comment]
@mgudemann said:
> Allows for bounding the unwinding of virtual functions in symex phase.
What does "bounding the unwinding of virtual functions" mean -- that is, what is meant by "unwinding" a function?

I can see what the code does, but it's not clear to me how this extra book keeping is beneficial. Assuming this is beneficial in some why, why does it apply to virtual functions only, and not to function pointers?

**Comment by mgudemann** [comment]
obsolete

**Pull request closed without merging**